### PR TITLE
test(client): initial Jest + RTL setup

### LINKED ISSUE #1937: Add tests 🧪
## Summary 💭

Currently, we don't have any test coverage. We rely on manual testing, Typescript, and inline code comments to ensure the library works as expected. 

## Description 📓

By adding unit/integration tests to the different modules exposed by the library: `next-auth/client,` `next-auth/providers`, etc... we will gain confidence during refactors and releases and better documentation than inline code comments. 💭 

The idea is to start testing our client module: `next-auth/client` as a POC to agree on the tooling and the workflows we want. Once we're happy with the setup, then go ahead and test the rest of the modules. 🏋🏽

### PR DESCRIPTION
**Pull request by 0ubbe**
## Reasoning 💡

This PR adds the initial unit/integration test setup with [Jest](https://jestjs.io/) + [RTL](https://testing-library.com/docs/react-testing-library/intro/) + [MSW](https://mswjs.io/). 

As a proof of concept, we have covered the existing acceptance criteria for `signIn()` and `getSession` from the **client** module: `next-auth/client`. The tests have been written in a fashion that mimics how users of this library will use these functions in the context o a React application.

Screenshot 2021-05-27 at 15 38 21

Screenshot 2021-05-27 at 15 35 03

I have also the tests to the CI, have a look at the [workflow](https://github.com/nextauthjs/next-auth/blob/5ceae67a3b8f7f1a4adff6132e153b7df74a3e5a/.github/workflows/tests.yml) (I feel we need to clean the actions soon) 🤔

...

## Checklist 🧢

- [ ] ~Documentation~
- [x] Tests
- [ ] ~Ready to be merged~
      

## Affected issues 🎟

Fixes #1937

### CODE REVIEW

**Review by ndom91 — APPROVED** [review]
Nice! LGTM :+1: 

I've always wanted to try out `msw`!

**Comment by ndom91** [comment]
Whats up with lots of these actions failing? They seem to be throwing errors about the `package-lock` not matching. 

I was about to push mine, but I see we're still using the `npm6` i.e. `lockfileVersion: 1`, so I couldn't as I only have npm7 available.

@lluia can you push your `package-lock.json` at some point as well? Thanks!

**Comment by ndom91** [comment]
@lluia Alright awesome! Good work!! 

I'm so happy that tests for this anddd the adapters are slowly but surely coming together 🍻🚀

File: package.json [hunk 1]
```diff
@@ -90,15 +90,20 @@
     "@babel/core": "^7.9.6",
     "@babel/plugin-proposal-optional-catch-binding": "^7.14.2",
     "@babel/plugin-transform-runtime": "^7.13.15",
-    "@babel/preset-env": "^7.14.2",
+    "@babel/preset-env": "^7.9.6",
```
Comment (lluia) [inline package.json:93]: Let's try to update all the `devDependencies` 👍🏽

File: package.json [hunk 2]
```diff
@@ -108,16 +113,21 @@
     "eslint-config-prettier": "^8.2.0",
     "eslint-config-standard-with-typescript": "^19.0.1",
     "eslint-plugin-import": "^2.22.1",
+    "eslint-plugin-jest": "^24.3.6",
     "eslint-plugin-node": "^11.1.0",
     "eslint-plugin-promise": "^4.3.1",
     "eslint-plugin-standard": "^5.0.0",
+    "jest": "^26.6.3",
+    "jest-mock-console": "^1.1.0",
```
Comment (balazsorban44) [inline package.json:121]: remove this

Comment (lluia) [inline package.json:121]: Remove this dependency 💅🏽

File: src/client/__tests__/mocks.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    return res(ctx.status(200), ctx.json(mockSession))
+  }),
+  rest.get("/api/auth/csrf", (req, res, ctx) => {
+    return res(ctx.status(200), ctx.json(mockCSRFToken))
+  }),
+  rest.get("/api/auth/providers", (req, res, ctx) => {
+    return res(ctx.status(200), ctx.json(mockProviders))
+  }),
+  rest.post("/api/auth/signin/github*", (req, res, ctx) => {
+    return res(ctx.status(200), ctx.json(mockGithubResponse))
```
Comment (lluia) [inline src/client/__tests__/mocks.js:69]: What about testing `500` and unhappy paths?

**Review by 0ubbe — COMMENTED** [review]
Few comments 💭

File: src/client/__tests__/sign-in.test.js [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+  userEvent.click(screen.getByRole("button"))
+
+  // if all went well, expect the user to be redirected to the provider post signin URL
+  await waitFor(() => {
+    expect(window.location.reload).toHaveBeenCalledTimes(1)
+    expect(window.location.replace).toHaveBeenCalledWith(mockUrlWithHash)
+  })
+})
+
+function SignInFlow({ providerId, callbackUrl, redirect = true }) {
```
Comment (lluia) [inline src/client/__tests__/sign-in.test.js:188]: Need to test as well `authorizationParams` option for `singIn()`

**Review by ndom91 — APPROVED** [review]
LGTM :+1:

**Pull request merged**